scipy/special/tests/test_support_cupy_torch_jax.py

```diff
@@ -0,0 +1,41 @@
+import pytest
+from hypothesis import given, strategies
+import hypothesis.extra.numpy as npst
+from numpy.testing import assert_allclose
+
+from scipy.special._support_cupy_torch_jax import (get_array_special_func,
+                                                   array_special_func_map)
+from scipy.conftest import array_api_compatible
+from scipy import special
+from scipy._lib.array_api_compat.array_api_compat import numpy as np
+
+
+def test_dispatch_failure():
+    message = ("SciPy cannot dispatch to special function 'ndtr' "
+               "in array library 'object'.")
+    with pytest.raises(ValueError, match=message):
+        get_array_special_func('ndtr', xp=object)
+
+
+@array_api_compatible
+@given(data=strategies.data())
+@pytest.mark.parametrize('f_name_n_args', array_special_func_map.items())
+def test_cupy_torch_jax_support(xp, data, f_name_n_args):
+    f_name, n_args = f_name_n_args
+    f = getattr(special, f_name)
+    dtypes = ['float32', 'float64']  # good enough for me
+    mbs = npst.mutually_broadcastable_shapes(num_shapes=n_args)
+
+    shapes, final_shape = data.draw(mbs)
+    dtype = data.draw(strategies.sampled_from(dtypes))
+    dtype = getattr(np, dtype)
+    args_np = [np.asarray(data.draw(npst.arrays(dtype, shape)))
+               for shape in shapes]
+    args_xp = [xp.asarray(arg) for arg in args_np]
+
+    ref = f(*args_np)
+    res = f(*args_xp)
+    # having trouble with this locally
+    assert type(res) == type(xp.asarray([]))
+    assert res.shape == ref.shape
+    assert_allclose(np.asarray(res), ref)
```

Comment:
Unless I am missing something, this `np.asarray` will fail for CuPy and PyTorch GPU. There has been discussion in #19005 of how best to go about testing on GPU. I have added this utility to `conftest.py`:
```python
def set_assert_allclose(xp=None):
    if xp is None:
        return npt.assert_allclose
    if 'cupy' in xp.__name__:
        return xp.testing.assert_allclose
    elif 'torch' in xp.__name__:
        return xp.testing.assert_close
    return npt.assert_allclose
```
This utility is called before every use of `assert_allclose`:
```python
 def test_ifft(self, xp):
        x = xp.asarray(random(30) + 1j*random(30))
        _assert_allclose = set_assert_allclose(xp)
        _assert_allclose(x, fft.ifft(fft.fft(x)))
```

Reply:
OK, I'll also implement this if gh-19005 merges first.
But what will be the canonical way to convert any Array API array to a different Array API array?

Reply:
As far as I can tell, this shouldn't be needed outside of tests. For tests, the use of `set_assert_allclose` is a deliberate attempt to avoid this as well.

The conversion to `np` and back is just so that we can make use of SciPy's existing functions, which only work for NumPy arrays. In general, there shouldn't be a need to convert between different array libraries.

Reply:
I think there was confusion here because I was referring to `np` as an Array API array.

You wrote
> Unless I am missing something, this `np.asarray` will fail for CuPy and PyTorch GPU

In that case, how *should* we be converting from any `xp` array to a `np` array outside of tests? (And until gh-19005 merges, I will do the conversion this way in tests before using `np.testing.assert_allclose`, too.)

Reply:
Ah, sorry for the confusion. We just use `np.asarray`, with `import numpy as np`. This is not the array API compatible version, but that doesn't matter since this is our 'last resort'. We expect it to raise exceptions, with GPU arrays for example. This can't be fixed without a general dispatch system, but that is out of scope for the array API work, hence your efforts to call to some particular libraries here.

Reply:
As for the tests, @tylerjereddy has had success using `array_api_compat.to_device` (in order to move GPU arrays to CPU before using `np.testing.assert_allclose`, however Ralf has expressed a preference for my method of `set_assert_allclose`, since `to_device` is not portable. The discussion of this happened quite recently on #19005.

Reply:
Sure, it's probably better to dispatch to the library's own testing function. 

OK, so I'm going to leave this line alone until gh-19005 merges (since CI doesn't test with GPU arrays) and in the `else` block of the special function wrapper, I'm using vanilla NumPy `np.asarray` as a last-ditch effort.

Reply:
I find it strange that there is no common way to convert between array types, though?

Reply:
Do you have an example of a conversion which you think should be possible? As far as I understand, the array API is supposed to be completely agnostic of the underlying library. Are you thinking of something like a general `to_numpy` function? This is something unsuitable for the array API, as it should hold up fine, regardless of the status of one particular project. I.e. forcing libraries to interact with other libraries in order to comply is undesirable.

Reply:
> Do you have an example of a conversion which you think should be possible?

I would expect `torch.asarray(np.asarray([1, 2, 3]))` to return a PyTorch array, and `np.asarray(torch.asarray([1, 2, 3]))` to return a NumPy array. Indeed, they do, so I would expect the same if we were to replace `np` and `torch` with any two Array libraries.

> Are you thinking of something like a general to_numpy function?

I was expecting that `np.asarray`, where `np` is an Array API compatible version of NumPy, would convert any Array API array to a NumPy array.

Reply:
Got you 👍. I am not so sure, but I believe that the blocker for this is device transfers. Imagine two array libraries which use different devices, both of which do not exist yet, but do in this hypothetical future.

How could we standardise the ability to transfer between these devices? It seems to me that you are hoping for the ability to transfer between any devices which could possibly be used by libraries which support the standard. Since the standard does not place any restrictions on the devices a library supports, I don't think that this i s possible.

I think that we're talking about the same thing here, but apologies if I've misunderstood.

cc @rgommers for clarification as I may also be misunderstanding the standard.

Reply:
> I would expect `torch.asarray(np.asarray([1, 2, 3]))` to return a PyTorch array, and `np.asarray(torch.asarray([1, 2, 3]))` to return a NumPy array

Yes indeed - when the data is on the same device.

> How could we standardise the ability to transfer between these devices? 

We can't really. At least, we haven't quite figured out how to best do this for testing purposes specifically. We do know that we never want to do this for actual library code.

Are we on the same page here now? The current PR diff says:
```
# TODO: use `set_assert_allclose` when gh-19005 merges
```
which sounds right (and if not, we'll have figured this out on gh-19005).

Reply:
> We do know that we never want to do this for actual library code.

You asked that the `else` case be to convert to NumPy, perform the computation, and convert back (https://github.com/scipy/scipy/pull/19023#discussion_r1287257353). So even for some Array API compatible arrays, this will just fail, and users will get whatever error `np.asarray` gives them?